toolchain/driver/driver_test.cpp

```diff
@@ -17,74 +23,108 @@
 namespace Carbon::Testing {
 namespace {
 
+using ::testing::ContainsRegex;
 using ::testing::ElementsAre;
 using ::testing::HasSubstr;
 using ::testing::StrEq;
 
+// Reads a file to string.
+// TODO: Extract this to a helper and share it with other tests.
+static auto ReadFile(std::filesystem::path path) -> std::string {
+  std::ifstream proto_file(path);
+  std::stringstream buffer;
+  buffer << proto_file.rdbuf();
+  proto_file.close();
+  return buffer.str();
+}
+
 class DriverTest : public testing::Test {
  protected:
-  DriverTest() : driver_(fs_, test_output_stream_, test_error_stream_) {}
+  DriverTest() : driver_(fs_, test_output_stream_, test_error_stream_) {
+    char* tmpdir_env = getenv("TEST_TMPDIR");
+    CARBON_CHECK(tmpdir_env != nullptr);
+    test_tmpdir_ = tmpdir_env;
+  }
 
-  auto CreateTestFile(llvm::StringRef text) -> llvm::StringRef {
-    static constexpr llvm::StringLiteral TestFileName = "test_file.carbon";
-    fs_.addFile(TestFileName, /*ModificationTime=*/0,
+  auto CreateTestFile(llvm::StringRef text,
+                      llvm::StringRef file_name = "test_file.carbon")
+      -> llvm::StringRef {
+    fs_.addFile(file_name, /*ModificationTime=*/0,
                 llvm::MemoryBuffer::getMemBuffer(text));
-    return TestFileName;
+    return file_name;
+  }
+
+  auto ScopedTempWorkingDir() {
```

Comment:
```suggestion
  // Makes a temp dir and sets the working directory. Returns a scope_exit that will delete it and restore the working directory.
  auto ScopedTempWorkingDir() {
```

I admit I'm a little uncomfortable with the auto return, although I assume it's because scope_exit is in llvm::detail. So I'd mainly suggest commenting to highlight the behavior.

Reply:
In addition to being in `detail`, I think it is also a template type that is hard to name. But definitely a good idea to comment this to clarify. I'll edit / enhance the comment a bit to have even more detail I think.